content/en/blog/_posts/2022-11-14-forensic-container-checkpointing/index.md

```diff
@@ -0,0 +1,207 @@
+during restore, I recommend that you use the latest version of CRI-O from the
+*main* CRI-O GitHub branch. If you're using CRI-O v1.25, you'll need to
+manually create certain directories Kubernetes would create before starting the
+container.
+
+The first step to restore a container outside of Kubernetes is to create a pod sandbox
+using *crictl*:
+```shell
+crictl runp pod-config.json
+```
+
+Then you can restore the previously checkpointed container into the newly created pod sandbox:
+```shell
+crictl create <POD_ID> container-config.json pod-config.json
+```
+
+Instead of specifying a container image in a registry in `container-config.json`
+you'd need specify the path to the checkpoint archive that you created earlier:
+```json
+{
+  "metadata": {
+      "name": "counter"
+  },
+  "image":{
+      "image": "/var/lib/kubelet/checkpoints/<checkpoint-archive>.tar"
+  }
+}
+```
+
+Next, run `crictl start <CONTAINER_ID>` to start that container, and then a
+copy of the previously checkpointed container should be running.
+
+### Restore a checkpointed container within of Kubernetes {#restore-checkpointed-container-k8s}
+
+To restore the previously checkpointed container directly in Kubernetes it is
+necessary to convert the checkpoint archive into an image that can be pushed to
+a registry.
+
+One possible way to convert the local checkpoint archive consists of the
+following steps with the help of [buildah](https://buildah.io/):
+```shell
+newcontainer=$(buildah from scratch)
+buildah add $newcontainer /var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar /
+buildah config --annotation=io.kubernetes.cri-o.annotations.checkpoint.name=<container-name> $newcontainer
+buildah commit $newcontainer checkpoint-image:latest
+```
+
+The resulting image is not standardized and only works in combination with CRI-O.
+Please consider this image format as pre-alpha. There are ongoing discussions
+to standardize the format of checkpoint images like this. Important to remember
+is that this not yet standardized image format only works if CRI-O has been started
+with `--enable-criu-support=true` and `--drop-infra-ctr=false`. The security implications
+of starting CRI-O with CRIU support are not yet clear and therefore the functionality
+as well as the image format should be used with care.
+
+Now, you'll need to push that image to a container image registry. For example:
+```shell
+podman push localhost/checkpoint-image:latest container-image-registry.example/user/checkpoint-image:latest
+```
+
+To restore this checkpoint image (`container-image-registry.example/user/checkpoint-image:latest`), the
+image needs to be listed in the specification for a Pod. Here's an example
+manifest:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  namePrefix: example-
+spec:
+  containers:
+  - name: <container-name>
+    image: container-image-registry.example/user/checkpoint-image:latest
+  nodeName: <destination-node>
```

Comment:
is the nodeName required?

Reply:
If it's omitted, you're going to have to copy the checkpoint data to all nodes before you make that Pod.